[AIRFLOW-5931] Use os.fork when appropriate to speed up task execution.

ashb opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] https://issues.apache.org/jira/browse/AIRFLOW-5931

### Description

- [x] Rather than running a fresh python interpreter which then has to re-load
  all of Airflow and its dependencies we should use os.fork when it is
  available/suitable which should speed up task running, espeically for
  short lived tasks.

  I've profiled this and it took the task duration (as measured by the
  `duration` column in the TI table) from an average of 14.063s down to
  just 0.932s!

  I _could_ make this change deeper and bypass the `CLIFactory`/go directly to `_run_raw_task`, but this makes the change the minimum needed to work.

### Tests

- [x] No unit tests added. Hopefully existing tests good enough. Manual testing shows this working

Other tests I need to perform:

- [x] Check if `os._exit` is right (this doesn't run atexit callbacks) - so I need to check if logging in the subprocess istidied up properly.
- [x] Test if this leaves "dangling"/broken DB connections.
- [x] Check remote log uploading

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

ashb committed:
[AIRFLOW-5931] Use os.fork when appropriate to speed up task execution.

Rather than running a fresh python interpreter which then has to re-load
all of Airflow and its dependencies we should use os.fork when it is
available/suitable which should speed up task running, espeically for
short lived tasks.

I've profiled this and it took the task duration (as measured by the
`duration` column in the TI table) from an average of 14.063s down to
just 0.932s!

kaxil commented on airflow/task/task_runner/standard_task_runner.py:
```diff
@@ -17,28 +17,69 @@
-            reap_process_group(self.process.pid, self.log)
+    def _start_by_fork(self):
+        pid = os.fork()
+        if pid:
```
Just wondering if we want to make it explicit. Maybe not required by just a suggestion:

```
if pid == 0:
```

example:


```python
    def _start_by_fork(self):
        pid = os.fork()
        if pid == 0:
            from airflow.bin.cli import CLIFactory
            parser = CLIFactory.get_parser()
            args = parser.parse_args(self._command[1:])
            setproctitle(
                "airflow task runner: {0.dag_id} {0.task_id} {0.execution_date} {0.job_id}".format(args)
            )
            args.func(args)
            os._exit(0)

        else:
            self.log.info("Started process %d to run task", pid)
            return psutil.Process(pid)
```

ashb replied:
Could do, but for python it doesn't matter so much - either it turns a pid, 0 in the child or throws an error (C can return 0, pid or -1 on error, but python converts that to an exception for us)

dimberman commented on airflow/task/task_runner/standard_task_runner.py:
```diff
@@ -17,28 +17,89 @@
-        if self.process and psutil.pid_exists(self.process.pid):
-            reap_process_group(self.process.pid, self.log)
+        if self.process:
+            if self.process.is_running():
```
can you break this into a function?

bolkedebruin commented on airflow/task/task_runner/standard_task_runner.py:
```diff
@@ -17,28 +17,89 @@
-            reap_process_group(self.process.pid, self.log)
+        if self.process:
+            if self.process.is_running():
+                reap_process_group(self.process.pid, self.log)
```
NIT: process groups are undeterministic. We would be better off using cgroups

ashb replied:
This is just so that `os.pgkill` beahves itself, not for anything more than that.

We've already got a Cgroup task runner (not that I know what state it is in!)

Errr. This is not the on the line I thought it was. Yea, pgroups might not be perfect but this isn't a change to the existing behaviour.

ashb committed:
Allow `reap_process_group` to kill processes even when the "group
leader" has already exited.

kaxil commented on airflow/task/task_runner/standard_task_runner.py:
```diff
@@ -17,28 +17,94 @@
-    def return_code(self):
-        return self.process.poll()
+            parser = get_parser()
+            args = parser.parse_args(self._command[1:])
```
Is this to remove `airflow` from the list?

ashb replied:
Yes.

kaxil reviewed (approved):
Looks Good.

potiuk commented on airflow/utils/helpers.py:
```diff
@@ -263,71 +263,82 @@ def f(t):
-    children = parent.children(recursive=True)
-    children.append(parent)
+    if pgid == os.getpgid(0):
+        raise RuntimeError("I refuse to kill myself")
```
Love it :)

mik-laj commented on airflow/cli/commands/task_command.py:
```diff
@@ -225,6 +226,11 @@ def task_test(args, dag=None):
         else:
             raise
+    finally:
+        if not already_has_stream_handler:
```
Thanks for it!

mik-laj commented on airflow/task/task_runner/standard_task_runner.py:
```diff
@@ -17,28 +17,96 @@
-    def return_code(self):
-        return self.process.poll()
+            parser = get_parser()
+            args = parser.parse_args(self._command[1:])
```
```suggestion
            # [1:] - remove "airflow" from the command
            args = parser.parse_args(self._command[1:])
```

mik-laj commented on tests/dags/test_on_kill.py:
```diff
@@ -25,6 +25,11 @@
 
 class DummyWithOnKill(DummyOperator):
     def execute(self, context):
+        import os
```
@nuclearpinguin Did you fix these issue in other PR?

ashb replied:
This wasn't an issue, I was just making doubley sure that the tests created more processes.

ashb committed:
Don't re-initialize JSON/stdout logging ElasticSearch inside forked processes

Most of the time we will run the "raw" task in a forked subprocess (the
only time we don't is when we use impersonation) that will have the
logging already configured. So if the EsTaskHandler has already been
configured we don't want to "re"configure it -- otherwise it will
disable JSON output for the actual task!
